File: agent/internal/cli/push_tree.go
// Copyright (c) 2025 kk
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

package cli

import (
	"fmt"

	"github.com/kk/kkartifact-agent/internal/client"
	"github.com/spf13/cobra"
)

var pushTreeCmd = &cobra.Command{
	Use:          "push-tree [root-path]",
	Short:        "Push all app/version directories under a root path",
	Long:         "Traverse root/{app}/{version}/ and push each version. Matches layout like /data/vcs/G02/tidb/G02_agent_api/<hash>/",
	SilenceUsage: true,
	RunE:         runPushTree,
}

var (
	pushTreeRoot         string
	pushTreeProject      string
	pushTreeConfig       string
	pushTreeServerURL    string
	pushTreeToken        string
	pushTreeConcurrency  int
	pushTreeIgnore       []string
	pushTreeDryRun       bool
	pushTreeSkipExisting bool
	pushTreePublish      bool
)

func init() {
	rootCmd.AddCommand(pushTreeCmd)

	pushTreeCmd.Flags().StringVar(&pushTreeRoot, "root", "", "Root directory containing app/version subdirectories")
	bindPushConfigFlags(pushTreeCmd, &pushTreeProject, &pushTreeConfig, &pushTreeServerURL, &pushTreeToken, &pushTreeConcurrency, &pushTreeIgnore)
	pushTreeCmd.Flags().BoolVar(&pushTreeDryRun, "dry-run", false, "Print planned pushes without uploading")
	pushTreeCmd.Flags().BoolVar(&pushTreeSkipExisting, "skip-existing", false, "Skip versions that already exist on the server")
	pushTreeCmd.Flags().BoolVar(&pushTreePublish, "publish", false, "Mark each version as published after push")
}

func runPushTree(cmd *cobra.Command, args []string) error {
	root := pushTreeRoot
	if len(args) > 0 {
		root = args[0]
	}
	if root == "" {
		return fmt.Errorf("root path is required (argument or --root)")
	}

	targets, err := listPushTreeTargets(root)
	if err != nil {
		return err
	}
	if len(targets) == 0 {
		fmt.Println("No app/version directories found")
		return nil
	}

	baseCfg, project, err := loadPushConfig(
		pushTreeProject,
		pushTreeConfig, pushTreeServerURL, pushTreeToken,
		pushTreeConcurrency, pushTreeIgnore,
	)
	if err != nil {
		return err
	}

	apiClient, err := client.New(baseCfg.ServerURL, baseCfg.Token)
	if err != nil {
		return fmt.Errorf("failed to create API client: %w", err)
	}

	fmt.Printf("Found %d version(s) under %s (project: %s)\n", len(targets), root, project)

	var failed int
	for _, target := range targets {
		params := &pushParams{
			project:   project,
			app:       target.app,
			version:   target.version,
			path:      target.path,
			cfg:       baseCfg,
			publish:   pushTreePublish,
			apiClient: apiClient,
		}

		label := fmt.Sprintf("%s/%s", params.app, params.version)
		if pushTreeDryRun {
			fmt.Printf("[dry-run] would push %s/%s:%s from %s\n", params.project, params.app, params.version, params.path)
			continue
		}

		if pushTreeSkipExisting && versionExistsOnServer(apiClient, params.project, params.app, params.version) {
			fmt.Printf(">>> skip existing %s\n", label)
			continue
		}

		fmt.Printf(">>> push %s\n", label)
		if err := executePush(params); err != nil {
			fmt.Printf("ERROR: %s: %v\n", label, err)
			failed++
		}
		fmt.Println()
	}

	if failed > 0 {
		return fmt.Errorf("%d push(es) failed", failed)
	}
	return nil
}

File: scripts/push-tree.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
# 批量 push：遍历 root/{app}/{version}/ 并调用 kkartifact-agent push-tree

require 'optparse'

options = {
  project: ENV['KK_PROJECT'],
  dry_run: false,
  skip_existing: false,
  publish: false,
  config: '.kkartifact.yml'
}

parser = OptionParser.new do |opts|
  opts.banner = 'Usage: ruby scripts/push-tree.rb [root-path] [options]'
  opts.on('-p', '--project PROJECT', '项目名称（或在配置文件中设置 project）') { |v| options[:project] = v }
  opts.on('--config PATH', 'Agent 配置文件路径') { |v| options[:config] = v }
  opts.on('--dry-run', '仅打印将要执行的 push') { options[:dry_run] = true }
  opts.on('--skip-existing', '跳过服务端已存在的版本') { options[:skip_existing] = true }
  opts.on('--publish', '推送成功后自动发布版本') { options[:publish] = true }
end
parser.parse!

root = ARGV[0] || '/data/vcs/G02/tidb'
abort("目录不存在: #{root}") unless File.directory?(root)

cmd = ['kkartifact-agent', 'push-tree', root]
cmd += ['--project', options[:project]] if options[:project]
cmd += ['--config', options[:config]]
cmd << '--dry-run' if options[:dry_run]
cmd << '--skip-existing' if options[:skip_existing]
cmd << '--publish' if options[:publish]

puts "执行: #{cmd.join(' ')}"
success = system(*cmd)
abort('push-tree 失败') unless success
